### DISCUSSION TITLE
$InSb_{x}As_{1-x}$ ternary alloy structure generation

### DISCUSSION THREAD

**Discussion by riCatwik** [pyiron/pyiron_atomistics#1021] [Q&A]
Dear all Pyiron developers, users and enthusiasts, 

I am new here. This is my first problem with Pyiron, but I am stuck at the very beginning. 
I would like to generate a ternary $InSb_xAs_{1-x}$ alloy. 

Atom positions of primitive zincblende InSb ($a=6.479 \unicode{x212B}$) are: 
```
0.00  0.00  0.00    In
0.25  0.25  0.25    Sb
```

I want an SQS structure in the following way: 
```
0.00  0.00  0.00    In
0.25  0.25  0.25    Sb=0.25 , As=0.75
```

Here I would like to 
- keep the positions of In atoms at (0 0 0) and its equivalent positions unchanged (or unperturbed - not sure what terminology to use) and 
- want only atoms at (0.25 0.25 0.25) and its equivalent positions to be randomized with Sb and As atoms 

The general command for SQS generation is "job.input.mole_fractions = {'A': 0.8, 'B':0.2}" but I am unsure how to do it for my specific purpose as I don't want to change the positions of In atoms. 

Could someone please help me to solve this? 
Looking forward to hearing from you. 

Thank you very much. 

Regards, 
Ritwik

**Comment by jan-janssen** [comment]
Hi @riCatwik , 

pyiron internally uses https://github.com/dgehringer/sqsgenerator from @dgehringer to generate SQS structures. If I understand your example correctly then you can use the input syntax discussed in:
https://sqsgenerator.readthedocs.io/en/latest/input_parameters.html#composition

But I agree it is not directly clear how to use it in pyiron. @dgehringer can you help us here? 

Best, 

Jan

**Reply by riCatwik** [reply]
Thanks for the information. 
Yes, I understand how it is done in `sqsgenerator`, but I still could not figure out how it can be done in Pyiron.

**Comment by samwaseda** [comment]
I know the answer and I can post it as soon as I can make my atomistics class run again (currently I’m suffering from the fact that my installation is not compatible)

**Reply by samwaseda** [reply]
I give you the outline here: First you can create a 2-atom structure (here I do it with Fe because I don’t know the base crystal structure of InSbAs):

```python
n_repeat = 3
bcc = pr.create.structure.bulk("Fe", cubic=True)
bcc[1] = "As"
structure = bcc.repeat(n_repeat)
```

And then you can run the SQS calculation only on the sublattice:

```python
sqs = pr.create.job.SQSJob("sqs_job")
sqs.input.mole_fractions = {"As": 0.75, "Sb": 0.25}
sqs.structure = structure[structure.select_index(“As”)]
sqs.run()
```

Then you can retrieve the chemical symbols via:

```python
chem_symbols = structure.get_chemical_symbols()
chem_symbols[structure.select_index(“As”)] = sqs[“I_cant_remember_the_tag”]
structure.set_chemical_symbols(chem_symbols)
```

Something like this should work. As I said I couldn’t test it myself because my installation is not working, but as soon as I manage to run it I will post the final version.

**Comment by riCatwik** [comment]
@jan-janssen @dgehringer 

Thank you very much, @samwaseda, for your answer. 

I followed your instruction and did the following: 

### 1. Structure generation
```python
from pyiron import Project
pr = Project(path="sqs")
```

```python
n_repeat  = 3
structure = pr.create.structure.bulk(name="InSb", 
                                     crystalstructure="zincblende", 
                                     a=6.479, 
                                     cubic=True)
structure = structure.repeat(n_repeat)
len(structure)
```

### 2. SQS generation on sublattice

```python
job_sqs = pr.create.job.SQSJob(job_name="sqs_job", 
                               delete_existing_job=True)
job_sqs.input.mole_fractions = {"Sb": 0.5, "As": 0.5}
job_sqs.input.iterations     = 1e6   ##  default:1000000 or 1e6
job_sqs.structure = structure[structure.select_index("Sb")]
job_sqs.run()
```

#### Print the first resulting SQS sub-structure

```python
structure_sublattice_sqs = job_sqs.list_structures()[0]
structure_sublattice_sqs
```

### 3. Retrieve chemical symbols and merge sublattice with main structure

```python
chem_symbols = structure.get_chemical_symbols()
chem_symbols[structure.select_index("Sb")] = structure_sublattice_sqs.get_chemical_symbols()
print(chem_symbols)

# Set the chemical symbols in the structure
structure_sqs = structure
structure_sqs.set_chemical_symbols(chem_symbols)

# Print the updated structure
structure_sqs
```

But, unfortunately, SQS structure symbols do not reflects in the main structure. 
When I did it with `ase` and `pyiron_atomistics` packages, it works. The code is below. 

```python
from pyiron_atomistics import ase_to_pyiron
from ase import Atoms

# Define the crystal structure
atoms = Atoms('In2Sb2As2',
              scaled_positions=[(0, 0, 0), (0.5, 0.5, 0), (0.5, 0, 0.5), (0, 0.5, 0.5)],
              cell=[6.479, 6.479, 6.479],
              pbc=True)

# Convert the ase Atoms object to a pyiron Atoms object
structure = ase_to_pyiron(atoms)

# Define the chemical symbols array with a data type of <U3
chem_symbols = np.array(['In', 'Sb', 'As', 'In', 'Sb', 'As'], dtype='<U3')

# Set the chemical symbols in the structure
structure.set_chemical_symbols(chem_symbols)

# Print the updated structure
print(structure)
```


So, I suppose there is something wrong with this line in `step 3`. 

```python
structure_sqs.set_chemical_symbols(chem_symbols)
```
I could not catch where it's going wrong. I also checked the data type of `chem_symbols` (`dtype='<U2'`) length of arrays, which all seems correct to me. 

Could you please help me to resolve this?

Thank you very much once again. 

Best, 
Ritwik

**Reply by samwaseda** [reply]
Ok I’m not surprised that it didn’t work because I also didn’t know exactly how to change the chemical symbols (which I would have obviously checked if I could have). Let me check what’s wrong in pyiron and I’ll come back to you. For now you can maybe use `pyiron_to_ase` to change the atom species as it’s done in ASE, I presume, right?
